File: 09.Greed贪心/leetcode_1833_maxIceCream.py
from typing import List

class Solution:
    def maxIceCream(self, costs: List[int], coins: int) -> int:
        mx = max(costs)
        tot = [0]*(mx+1)
        for i in costs:
            tot[i] += 1
        
        res = 0
        for i in range(1, mx+1):
            if coins < i:
                break
            cnt = min(coins//i, tot[i])
            res += cnt
            coins -= i*cnt
        return res

if __name__ == "__main__":
    s = Solution()
    costs = [1,3,2,4,1]
    coins = 7
    print(s.maxIceCream(costs, coins))

File: 09.Greed贪心/leetcode_1833_maxIceCream.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxIceCream(vector<int>& costs, int coins) {
        int mx = *max_element(costs.begin(), costs.end());
        vector<int> tot(mx+1, 0);
        for (int i : costs){
            tot[i]++;
        }

        int res = 0;
        for (int i=1; i<=mx; i++){
            if (coins < i){
                break;
            }
            int cnt = min(coins/i, tot[i]);
            coins -= i*cnt;
            res += cnt;
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<int> costs = {1,3,2,4,1};
    int coins = 7;
    cout << s.maxIceCream(costs, coins) << endl;
}
